### ISSUE TITLE
Tuning the batch size in the Tuner component

### ISSUE THREAD

**Issue by jimzer** [tensorflow/tfx#3591]
Hello,

I'm currently using TFX to build a pipeline on the Google AI platform with the Kubeflow engine. I have a model where the **batch size** is an important hyper-parameter to tune.

I would like to search this hyper-parameter in the **Tuner component**.

Is it even possible?

I follow the TFX example with the Penguin dataset, more precisely the tuner component implementation:[ found here](https://github.com/tensorflow/tfx/blob/dd9829b0ac6e2dd708641ce39e7a3234a552d56b/tfx/examples/penguin/penguin_utils_cloud_tuner.py).

The `_get_hyperparameters` function returns the sample space for the model hyperparameters ([see line 139](https://github.com/tensorflow/tfx/blob/dd9829b0ac6e2dd708641ce39e7a3234a552d56b/tfx/examples/penguin/penguin_utils_cloud_tuner.py#L139)). However, the batch size to train the model is fixed and specified at the end of the tuner_fn ([see line 246](https://github.com/tensorflow/tfx/blob/dd9829b0ac6e2dd708641ce39e7a3234a552d56b/tfx/examples/penguin/penguin_utils_cloud_tuner.py#L246)).

Is there a way to **dynamically** change the **batch size**  based on a sample from the hyper-parameter space?

Thanks for your help !

**Comment by 1025KB** [comment]
Hi, 

The Tuner currently only support tuning the hparams of the model for best metrics (e.g., accuracy)

batch_size is outside of the model tuning loop, and it's for performance instead of accuracy. Thus the existing tuning logic (depends on kerastuner library) won't be able to handle it.

(batch_size is more like a TFX pipeline configuration tuning loop instead of model hparam tuning loop)

**Comment by 1025KB** [comment]
Customize kerastuner.BaseTuner should work for batch size tuning

you need to create a XXXTuner with a [run_trial](https://github.com/keras-team/keras-tuner/blob/master/kerastuner/engine/tuner.py#L150)() that:
1. get the batch_size from trial
2. create input with correct batch_size
3. train the model with created input
4. you might need to update how the metrics is calculated if you want to add execution time as one of the evaluation method for trial

And in your tuner_fn you need to change the [fit_args](https://github.com/tensorflow/tfx/blob/master/tfx/examples/penguin/penguin_utils_keras.py#L122) to match the run_trial params

**Comment by jimzer** [comment]
Thanks for your help!
I will give a try to the Keras BaseTuner customization.

**Comment by arghyaganguly** [comment]
@jimzer ,please close this if you are satisfied with the solution provided by @1025KB.Thanks.
